### PR TITLE
Add `TemplateNode` and `HTMLNode` subclasses

### PR DESCRIPTION
**Pull request by xxyzz**
The parse is broken for the two HTML tests...

### CODE REVIEW

File: wikitextprocessor/parser.py [hunk 1]
```diff
@@ -334,11 +334,40 @@ def __str__(self) -> str:
...
+        for node in self.find_child_recursively(target_kind):
+            return True
+        return False
 
 class TemplateNode(WikiNode):
     def __init__(self, linenum: int):
         super().__init__(NodeKind.TEMPLATE, linenum)
-        self._template_parameters: Optional[Dict[str | int, str | WikiNode]] = None
+        self._template_parameters: Optional[
+            Dict[str | int, str | WikiNode]
```
Comment (kristian-clausal) [inline wikitextprocessor/parser.py:380]: We haven't yet upgraded to a version of Python that support the pipe for type annotations. Please use Union[]

Comment (xxyzz) [reply wikitextprocessor/parser.py:380]: I forget to use Union... Actually we can use `|`, it was introduced in Python 3.10.

**Comment by xxyzz** [comment]
> There are a few places in the code where the type of node a WikiNode has is changed through assigning a different NodeKind to it, just to keep an eye out.

I didn't know that, will search the code to see how the node type is changed.

And the html test returns strange tree, I'm not sure if this is expected:

The parse return of `<h1><p></p></h1>` is `<ROOT(['t']){} <HTML(h1){} >, <HTML(p){} >, '</h1>'>`, this doesn't look right.

**Comment by kristian-clausal** [comment]
We have 3.8 here, so that's what we're targeting. The newest Python available on certain distros. On Aug 24, 2023 15:52, xxyzz ***@***.***> wrote:
@xxyzz commented on this pull request.



In wikitextprocessor/parser.py:
>  
 class TemplateNode(WikiNode):
     def __init__(self, linenum: int):
         super().__init__(NodeKind.TEMPLATE, linenum)
-        self._template_parameters: Optional[Dict[str | int, str | WikiNode]] = None
+        self._template_parameters: Optional[
+            Dict[str | int, str | WikiNode]

I forget to use Union... Actually we can use |, it was introduced in Python 3.10.

File: wikitextprocessor/parser.py (hunk 1 above)
Comment (xxyzz) [reply wikitextprocessor/parser.py:380]: Maybe we could consider using [container](https://podman.io) or [pyenv](https://github.com/pyenv/pyenv) on the server, using Python 3.11 is the easiest way to improve speed: https://docs.python.org/3/whatsnew/3.11.html#faster-cpython

**Comment by xxyzz** [comment]
I find only two lines of code set the node type to parser function:

https://github.com/tatuylonen/wikitextprocessor/blob/e5296c16f2d715e62121f23cb5057374da48cda3/wikitextprocessor/parser.py#L430

https://github.com/tatuylonen/wikitextprocessor/blob/e5296c16f2d715e62121f23cb5057374da48cda3/wikitextprocessor/parser.py#L919

File: wikitextprocessor/parser.py (hunk 1 above)
Comment (kristian-clausal) [reply wikitextprocessor/parser.py]: I was wrong when I said the dev machine was using Python 3.8, that was my own local machine... The dev is on 3.10.

However, it is unlikely we will upgrade to 3.11 any time soon.

**Comment by xxyzz** [comment]
Even though the above two lines change template kind to parser function, the extractor code still uses `isinstance(node, WikiNode)` and  `node.kind` to check the node type but not the class name, so I think this won't be a problem.

**Comment by kristian-clausal** [comment]
At some point, we probably want to get rid of the node.kind stuff and only have classes, but yeah, in the meanwhile node.kind is fine.

File: wikitextprocessor/core.py [hunk 2]
```diff
@@ -1692,37 +1692,46 @@ def get_page(
...
+                # Wikitext and Lua code
                 if title.startswith(ns_prefix):
                     template_name = title[len(ns_prefix) :]
-                    title = (
+                    lower_case_title = (
                         ns_prefix + template_name[0].upper() + template_name[1:]
                     )
                 else:
-                    title = ns_prefix + title[0].upper() + title[1:]
+                    lower_case_title = ns_prefix + title[0].upper() + title[1:]
```
Comment (kristian-clausal) [inline wikitextprocessor/core.py:1703]: If you use .upper() here, then `lower_case_title` has the wrong name; it's an upper_case_title.

Comment (xxyzz) [reply wikitextprocessor/core.py:1703]: Thanks! I don't know why I named it `lower_case`...

**Comment by xxyzz** [comment]
I'm trying to use the new class method and attribute in wiktextract and find the code at here also create `WikiNode`: https://github.com/tatuylonen/wiktextract/blob/205c4a2d88c27113f0117e0095f466605976af81/wiktextract/page.py#L80

**Pull request merged**